src/main/java/com/coveo/pushapiclient/DocumentUploadQueue.java

```diff
@@ -1,21 +1,105 @@
 package com.coveo.pushapiclient;
 
 import java.io.IOException;
+import java.util.ArrayList;
 
-// TODO: LENS-851 - Make public
+/**
+ * Represents a queue for uploading documents using a specified upload strategy
+ */
 class DocumentUploadQueue {
     private final UploadStrategy uploader;
+    private final int maxQueueSize = 5 * 1024 * 1024;
+    private ArrayList<DocumentBuilder> documentToAddList;
+    private ArrayList<DeleteDocument> documentToDeleteList;
+    private int size;
 
+    /**
+     * Constructs a new DocumentUploadQueue object with a default maximum queue size
+     * limit of 5MB.
+     *
+     * @param uploader The upload strategy to be used for document uploads.
+     */
     public DocumentUploadQueue(UploadStrategy uploader) {
+        this.documentToAddList = new ArrayList<>();
+        this.documentToDeleteList = new ArrayList<>();
         this.uploader = uploader;
     }
 
+    /**
+     * Flushes the accumulated documents by applying the upload strategy.
+     *
+     * @throws IOException          If an I/O error occurs during the upload.
+     * @throws InterruptedException If the upload process is interrupted.
+     */
     public void flush() throws IOException, InterruptedException {
-        throw new UnsupportedOperationException("Unimplemented method (TODO: LENS-856)");
+        if (this.isEmpty()) {
+            return;
+        }
+        BatchUpdate batch = this.getBatch();
+        // TODO: LENS-871: support concurrent requests
+        this.uploader.apply(batch);
+        this.size = 0;
+        this.documentToAddList.clear();
+        this.documentToDeleteList.clear();
     }
 
+    /**
+     * Adds a {@link DocumentBuilder} to the upload queue and flushes the queue if
+     * it exceeds the maximum content length.
+     * See {@link DocumentUploadQueue#flush}.
+     *
+     * @param document The document to be added to the index.
+     * @throws IOException          If an I/O error occurs during the upload.
+     * @throws InterruptedException If the upload process is interrupted.
+     */
     public void add(DocumentBuilder document) throws IOException, InterruptedException {
-        throw new UnsupportedOperationException("Unimplemented method (TODO: LENS-856)");
+        if (document == null) {
+            return;
+        }
+
+        final int sizeOfDoc = document.marshal().getBytes().length;
```

Comment:
Marshalling documents twice (once to measure size, a second time to upload) seems wasteful, but I don't really know a better solution since at this point we're still taking a document builder as a parameter and BatchUpdate exposes both its lists publicly.

Reply:
I know 😢 
But like you said, that is a side effect of accepting DocumentBuilder instead of stringified JSON.

At this stage, we need to choose the lesser of two evils. Accept stringified JSON for better performance but no validation or DocumentBuilder for better UX/less performance.

Another solution:
The DocumentBuilder would keep a state of its size every time it gets modified. That way, we could the DocumentUploadQueue could simple do `document.getSize()`.
However, I feel that it would not be as accurate than marshalling first.

I am open to suggestions